src/libappimage/core/impl/TraversalType2.cpp

```diff
@@ -0,0 +1,333 @@
+
+    if (fs_offset < 0)
+        throw IOError("get_elf_size error");
+
+    sqfs_err err = sqfs_open_image(&fs, path.c_str(), (size_t) fs_offset);
+
+    if (err != SQFS_OK)
+        throw IOError("sqfs_open_image error: " + path);
+
+    // prepare for traverse
+    rootInodeId = sqfs_inode_root(&fs);
+    err = sqfs_traverse_open(&trv, &fs, rootInodeId);
+    if (err != SQFS_OK) {
+        sqfs_destroy(&fs);
+        throw IOError("sqfs_traverse_open error");
+    }
+
+    // Fetch first entry
+    next();
+}
+
+TraversalType2::~TraversalType2() {
+    sqfs_traverse_close(&trv);
+
+    clog << "Closing " << path << " as Type 2 AppImage" << endl;
+    sqfs_destroy(&fs);
+}
+
+void TraversalType2::next() {
+    sqfs_err err;
+    if (!sqfs_traverse_next(&trv, &err))
+        completed = true;
+
+    if (err != SQFS_OK)
+        throw IOError("sqfs_traverse_next error");
+
+    if (!completed) {
+        currentEntryType = readEntryType();
+        currentEntryPath = readEntryName();
+        currentEntryLink = currentEntryType == PayloadEntryType::LINK ? readEntryLink() : std::string();
+    } else {
+        currentEntryType = PayloadEntryType::UNKNOWN;
+        currentEntryPath = std::string();
+        currentEntryLink = std::string();
+    }
+}
+
+bool TraversalType2::isCompleted() const {
+    return completed;
+}
+
+std::string TraversalType2::getEntryName() const {
+    return currentEntryPath;
+}
+
+appimage::core::PayloadEntryType TraversalType2::getEntryType() const {
+    return currentEntryType;
+}
+
+void TraversalType2::extract(const std::string& target) {
+    sqfs_inode inode;
+    if (sqfs_inode_get(&fs, &inode, trv.entry.inode))
+        throw IOError("sqfs_inode_get error");
+
+    // create target parent dir
+    auto parentPath = bf::path(target).parent_path();
+    bf::create_directories(parentPath);
+
+    // handle each inode type properly
+    switch (inode.base.inode_type) {
+        case SQUASHFS_DIR_TYPE:
+        case SQUASHFS_LDIR_TYPE:
+            extractDir(target);
+            break;
+        case SQUASHFS_REG_TYPE:
+        case SQUASHFS_LREG_TYPE:
+            extractFile(inode, target);
+            break;
+        case SQUASHFS_SYMLINK_TYPE:
+        case SQUASHFS_LSYMLINK_TYPE:
+            extractSymlink(inode, target);
+            break;
+        default:
+            throw AppImageError("AppImage Type 2 inode.base.inode_type " +
+                                std::to_string(inode.base.inode_type) + " not supported yet");
+    }
+}
+
+sqfs_err TraversalType2::sqfsStat(sqfs* fs, sqfs_inode* inode, struct stat* st) {
+    // code borrowed from the "extract.c" file at squashfuse
+    sqfs_err err;
+    uid_t id;
+
+    memset(st, 0, sizeof(*st));
+    // fill stats
+    st->st_mode = inode->base.mode;
+    st->st_nlink = inode->nlink;
+    st->st_mtime = st->st_ctime = st->st_atime = inode->base.mtime;
+
+    if (S_ISREG(st->st_mode)) {
+        /* FIXME: do symlinks, dirs, etc have a size? */
```

Comment:
Erm... can you answer this question perhaps? FIXMEs don't look good in prod code.

Reply:
This is part of the code that I copied from the `squashfs` implementation. And the answer is yes. Removing that comment.

Reply:
This section of code was removed.